--- ../Rack/src/widget/OpenGlWidget.cpp	2022-11-23 23:11:38.000000000 +0000
+++ OpenGlWidget.cpp	2022-11-23 23:06:41.000000000 +0000
@@ -1,3 +1,30 @@
+/*
+ * DISTRHO Cardinal Plugin
+ * Copyright (C) 2021-2022 Filipe Coelho <falktx@falktx.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 3 of
+ * the License, or any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * For a full copy of the GNU General Public License see the LICENSE file.
+ */
+
+/**
+ * This file is an edited version of VCVRack's OpenGlWidget.cpp
+ * Copyright (C) 2016-2021 VCV.
+ *
+ * This program is free software: you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 3 of
+ * the License, or (at your option) any later version.
+ */
+
 #include <widget/OpenGlWidget.hpp>
 #include <context.hpp>
 
@@ -14,23 +41,6 @@
 
 
 void OpenGlWidget::drawFramebuffer() {
-	math::Vec fbSize = getFramebufferSize();
-	glViewport(0.0, 0.0, fbSize.x, fbSize.y);
-	glClearColor(0.0, 0.0, 0.0, 1.0);
-	glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT | GL_STENCIL_BUFFER_BIT);
-
-	glMatrixMode(GL_PROJECTION);
-	glLoadIdentity();
-	glOrtho(0.0, fbSize.x, 0.0, fbSize.y, -1.0, 1.0);
-
-	glBegin(GL_TRIANGLES);
-	glColor3f(1, 0, 0);
-	glVertex3f(0, 0, 0);
-	glColor3f(0, 1, 0);
-	glVertex3f(fbSize.x, 0, 0);
-	glColor3f(0, 0, 1);
-	glVertex3f(0, fbSize.y, 0);
-	glEnd();
 }
 
 
